### PR TITLE
HW driver: drv2605l

### PR DESCRIPTION
**Pull request by jacobrosenthal**
Work in progress, but working drv2605l (mostly lra) driver implementation.

Any comments appreciated.

### CODE REVIEW

**Comment by jacobrosenthal** [comment]
For anyone here who might have a drv2605 and an LRA lying around, heres a starting point and something that can be a tutorial soon:


```
add the driver in your target's pkg.yml
pkg.deps:
  "@apache-mynewt-core/hw/drivers/drv2605"

Turn on everything in your syscfg.yml
    I2C_0: 1 #drv2605
    DRV2605_OFB: 1
    DRV2605_CLI: 1
    DRV2605_LOG: 1
    DRV2605_STATS: 1
    
    
and now the hard part you need to read the drv2605l datasheet or get a bunch of numbers from your motor manufacturer so you can override your motors voltage and other settings. 
    # these are settings taken from the drv2605 evm for the SEMCO1030 (Mplus)
    # 205hz freq, nominal voltage 1.5V, max voltage 3.0V, DRV2605_LRA_SAMPLE_TIME: 3, DRV2605_LRA_BLANKING_TIME: 2, DRV2605_LRA_IDISS_TIME: 2

    DRV2605_RATED_VOLTAGE: 0x3E # See 8.5.2.1 Rated Voltage Programming
    DRV2605_OD_CLAMP: 0x89 # '8.5.2.2 Overdrive Voltage-Clamp Programming, Note: LRA and ERM (8) and (9) equations are swapped in the "SLOS854C REVISED SEPTEMBER 2014" datasheet, confirmed https://e2e.ti.com/support/other_analog/haptics/f/927/t/655886'
    DRV2605_DRIVE_TIME: 19 # '8.5.2.2 Overdrive Voltage-Clamp Programming, Note: LRA and ERM (8) and (9) equations are swapped in the "SLOS854C REVISED SEPTEMBER 2014" datasheet, confirmed https://e2e.ti.com/support/other_analog/haptics/f/927/t/655886'
    
    
And now you need calibration values, you might have gotten these from someone as well and you can just override them as well. These are the SEMCO1030 again. If you dont have them you're ok for now but youll need to generate them later

    DRV2605_CALIBRATED_COMP: 0x09
    DRV2605_CALIBRATED_BEMF: 0x79
    DRV2605_CALIBRATED_BEMF_GAIN: 1
        
Start up your board and head to the shell we turned on earlier

 
I recommend first things first see if you can talk to device with
053273 compat> drv2605 chip_id
drv2605 chip_id
118613 0x07

next see if your calibration numbers work (make sure your motor is secured to something)
118613 compat> drv2605 diag
drv2605 diag
120858 Diagnostic Complete


If they didnt work or you didnt have good values overriden from before youll need to generate them
if you need to autocal your device you can do that from here too with autocal brake_factor loop_gain lra_sample_time lra_blanking_time lra_idiss_time auto_cal_time lra_zc_det_time
NOTE you need to figure out your numbers, not use these
001407 compat> drv2605 autocal 3 1 2 3 2 2 0
drv2605 autocal 3 1 2 3 2 2 0
001931 Autocal Complete
001931 DRV2605_CALIBRATED_COMP: 0x2C
DRV2605_CALIBRATED_BEMF: 0x49
DRV2605_CALIBRATED_BEMF_GAIN: 3

now youd want to save these as overrides in your target's syscfg.yml

next try to load some of the internal roms, 4 hard clicks (1) with max delays in between (255)
120858 compat> drv2605 load 1 255 1 255 1 255 1 255
drv2605 load 1 255 1 255 1 255 1 255
122555 Load succeeded

now you can trigger those forms
122555 compat> drv2605 trigger
drv2605 trigger
128806 Trigger succeeded

Its handy to dump the register contents as you're poking around to see if what you think you did worked
001931 compat> drv2605 dump_all
drv2605 dump_all
069774 reg 0x:00 = 0xE0
069774 reg 0x:01 = 0x40
069774 reg 0x:02 = 0x00
069774 reg 0x:03 = 0x06
069774 reg 0x:04 = 0x01
069774 reg 0x:05 = 0xFF
069774 reg 0x:06 = 0x01
069774 reg 0x:07 = 0xFF
069775 reg 0x:08 = 0x01
069775 reg 0x:09 = 0xFF
069775 reg 0x:0A = 0x01
069775 reg 0x:0B = 0xFF
069775 reg 0x:0C = 0x00
069775 reg 0x:0D = 0x00
069775 reg 0x:0E = 0x00
069775 reg 0x:0F = 0x00
069775 reg 0x:10 = 0x00
069775 reg 0x:11 = 0x05
069775 reg 0x:12 = 0x19
069775 reg 0x:13 = 0xFF
069775 reg 0x:14 = 0x19
069775 reg 0x:15 = 0xFF
069775 reg 0x:16 = 0x3E
069775 reg 0x:17 = 0x89
069775 reg 0x:18 = 0x2C
069775 reg 0x:19 = 0x49
069775 reg 0x:1A = 0xB7
069775 reg 0x:1B = 0x93
069775 reg 0x:1C = 0x2E
069775 reg 0x:1D = 0x00
069775 reg 0x:1E = 0x20
069775 reg 0x:1F = 0x00
069776 reg 0x:20 = 0x33
069776 reg 0x:21 = 0x6F
069776 reg 0x:22 = 0x3D
069776 compat> 

you can also peek and poke individual registers, ill manually take it out of standby
074759 compat> drv2605 poke 1 0
drv2605 poke 1 0
083307 Wrote: 0x00 to 0x01
083307 compat> drv2605 peek 1
drv2605 peek 1
083883 Value: 0x00
```

**Comment by jacobrosenthal** [comment]
note DRV2605_OFB doesnt do anything, not sure where to put that since im not a sensor..
need this somewhere
```
#if MYNEWT_VAL(DRV2605_OFB)
#include "hal/hal_gpio.h"
#include <drv2605/drv2605.h>
static struct drv2605 drv2605;
#endif

#if MYNEWT_VAL(DRV2605_OFB)
static struct sensor_itf i2c_0_itf_drv = {
    .si_type = 1,
    .si_num  = 0,
    .si_addr = MYNEWT_VAL(DRV2605_SHELL_ITF_ADDR)
};
#endif

#if MYNEWT_VAL(DRV2605_OFB)
    rc = hal_gpio_init_out(30, 1);
    rc = os_dev_create((struct os_dev *) &drv2605, "drv2605_0",
      OS_DEV_INIT_PRIMARY, 0, drv2605_init, (void *)&i2c_0_itf_drv);
    assert(rc == 0);
#endif
```
also kinda wierd, but because its not a sensor nobody calls my drv2605_config, so im callilng it myself inside of my drv2605_init

**Comment by microbuilder** [comment]
Hi Jacob. @sterlinghughes asked me to have a look at this, but good to have a driver for this IC since it's a really nice chip, and a good match for BLE!

I only have the [Adafruit DRV2605 Breakout](https://www.adafruit.com/product/2305) and [this small motor](https://www.adafruit.com/product/1201) to test with, but I'll see what that gives me.

I did have a look at the code on my phone late last night, and it looks like some of the doxygen type comment header blocks on functions are copy/pasted and might mislead people in any generated documentation since they repeat the same descriptions, but I'll wait until I can test the code out on real HW before seriously digging into it.

Thanks for submitting this and getting a useful driver started for this chip, though.

**Comment by jacobrosenthal** [comment]
Yep, Ive started a bit of commenting on this other branch where Im playing with making it more newty and by using a runtime config for mode selection and seperate power mode setter/getter

https://github.com/jacobrosenthal/mynewt-core/tree/drv2605-config

Not actually guaranteed to work yet and not in line with above documentation.

**Comment by jacobrosenthal** [comment]
OK. So that branch is mostly working and Ill probably bring it in unless anyone has objections to the way I solved most of my existing questions. To test new branch youd:
* config rom mode (automatically setting en high and besides diag and cal, ends up with you in standby)
* set active
* load some rom selections
* trigger the playback
```
001414 compat> drv2605 op_mode rom
drv2605 op_mode rom
002111 op_mode succeeded
002111 compat> drv2605 power_mode active
drv2605 power_mode active
003263 power_mode succeeded
003263 compat> drv2605 load 1 255 1 255 1 255 1 255
drv2605 load 1 255 1 255 1 255 1 255
004939 load succeeded
004939 compat> drv2605 trigger
drv2605 trigger
005707 trigger succeeded
```

**Comment by jacobrosenthal** [comment]
Note Im hacking cs pin as enable. Youll need that set. Havent decided how to deal with en tied high, can we -1 a pin?

```
#if MYNEWT_VAL(DRV2605_OFB)
static struct sensor_itf i2c_0_itf_drv = {
    .si_type = 1,
    .si_num  = 0,
    .si_addr = MYNEWT_VAL(DRV2605_SHELL_ITF_ADDR),
    .si_cs_pin = 3
};
#endif
```

**Comment by jacobrosenthal** [comment]
And heres a programmatic example for the new branch that runs the same 4 beat pattern infinitely
```cpp
#include <drv2605/drv2605.h>
static struct os_callout motor_callout;

int
infinite_buzz(struct sensor_itf *itf)
{
    int rc;

    rc = drv2605_trigger_rom(itf);
    if(rc){
        return rc;
    }else{
        return os_callout_reset(&motor_callout, OS_TICKS_PER_SEC/20);
    }
}


static void
timer_ev_cb(struct os_event *ev)
{
    assert(ev != NULL);
    int rc;
    bool is_busy;
    struct sensor_itf *itf;
    
    itf = ev->ev_arg;

    //is this context safe? should this be os_eventq_put instead?
    rc = drv2605_rom_busy(itf, &is_busy);
    if(rc) {
        //fail silently I guess
    }else{
        if(is_busy){
            //50ms
            os_callout_reset(&motor_callout, OS_TICKS_PER_SEC/20);
        }else{
            infinite_buzz(itf);
        }
    }

}

int
main(int argc, char **argv)
{
...
    int rc;
    struct os_dev *dev;
    struct drv2605 *drv2605;
    struct drv2605_cfg cfg;

    dev = (struct os_dev *) os_dev_open("drv2605_0", OS_TIMEOUT_NEVER, NULL);
    assert(dev != NULL);

    memset(&cfg, 0, sizeof(cfg));

    cfg.op_mode = DRV2605_OP_ROM;

    rc = drv2605_config((struct drv2605 *)dev, &cfg);
    assert(!rc);

    drv2605 = (struct drv2605 *) dev;
    struct sensor_itf *itf;
    itf = SENSOR_GET_ITF(&(drv2605->sensor));

    rc = drv2605_set_power_mode(itf, DRV2605_POWER_ACTIVE);
    assert(!rc);

    uint8_t waveform[] = {1, 255, 1, 255, 1, 255, 1, 255};
    rc = drv2605_load_rom(itf, waveform, 8);
    assert(!rc);


    os_callout_init(&motor_callout, os_eventq_dflt_get(),
                    timer_ev_cb, itf);

    rc = infinite_buzz(itf);
...
}
```

**Comment by jacobrosenthal** [comment]
Merged that branch here.  Also just put everything in sensor creator so theres an obvious example, we can just rebase that out if people dont like it
For anyone here who might have a drv2605 and an LRA lying around, heres a starting point and something that can be a tutorial soon:

    
My syscfg. This is the hardest part as need to read the drv2605l datasheet or get a bunch of numbers from your motor manufacturer so you can override your motors voltage and other settings. 
```
    I2C_0: 1 #drv2605
    DRV2605_OFB: 1
    DRV2605_EN_PIN: 30
    DRV2605_CLI: 1
    DRV2605_LOG: 1
    DRV2605_STATS: 1

    #jinlong LVM061930 2.0v 2.35max 180hz
    DRV2605_RATED_VOLTAGE: 0x53 # See 8.5.2.1 Rated Voltage Programming
    DRV2605_OD_CLAMP: 0x69 # '8.5.2.2 Overdrive Voltage-Clamp Programming, Note: LRA and ERM (8) and (9) equations are swapped in the "SLOS854C REVISED SEPTEMBER 2014" datasheet, confirmed https://e2e.ti.com/support/other_analog/haptics/f/927/t/655886'
    DRV2605_DRIVE_TIME: 23 #8.6.21 LRA DRIVE_TIME = (0.5 × (1/fLRA * 1000) - 0.5 ms) /  0.1 ms
```
Ill use the sensors_test app, my target.yml:
```
### Target: targets/sensors-thingy
target.app: "@apache-mynewt-core/apps/sensors_test"
target.bsp: "@apache-mynewt-core/hw/bsp/nrf52-thingy"
target.build_profile: "optimized"
```

Add the driver in your target's pkg.yml, though note this is already done for the sensor_test example
```
pkg.deps:
  "@apache-mynewt-core/hw/drivers/drv2605"
```

Run the app with perhaps `newt run sensors-thingy 0`

Connect to your shell. I recommend first things first see if you can talk to device with
```
053273 compat> drv2605 chip_id
drv2605 chip_id
118613 0x07
```
next see if your motor numbers work (make sure your motor is secured to something, this will fail otherwise)
```
118613 compat> drv2605 diag
drv2605 diag
120858 Diagnostic Complete
```
If that didnt work or you didnt have good values overriden from before youll need to generate them
if you need to autocal your device you can do that from here too with autocal and no params to get the default best guess
```
001407 compat> drv2605 autocal
drv2605 autocal
001931 Autocal Complete
001931 DRV2605_CALIBRATED_COMP: 0x2C
DRV2605_CALIBRATED_BEMF: 0x49
DRV2605_CALIBRATED_BEMF_GAIN: 3
```
now youd want to save these as overrides in your target's syscfg.yml and restart

Now you're ready to (sigh) rumble Load some of the internal roms, 4 hard clicks (1) with max delays in between (255)
```
120858 compat> drv2605 load 1 255 1 255 1 255 1 255
drv2605 load 1 255 1 255 1 255 1 255
122555 Load succeeded
```
The motor is in standby by default so enable it
```
002111 compat> drv2605 power_mode active
drv2605 power_mode active
003263 power_mode succeeded
```
now you can trigger those forms, as many times as you want, or load new forms and trigger again.
```
122555 compat> drv2605 trigger
drv2605 trigger
128806 Trigger succeeded
```
Its handy to dump the register contents as you're poking around to see if what you think you did worked
```
001931 compat> drv2605 dump_all
drv2605 dump_all
069774 reg 0x:00 = 0xE0
069774 reg 0x:01 = 0x40
069774 reg 0x:02 = 0x00
069774 reg 0x:03 = 0x06
069774 reg 0x:04 = 0x01
069774 reg 0x:05 = 0xFF
069774 reg 0x:06 = 0x01
069774 reg 0x:07 = 0xFF
069775 reg 0x:08 = 0x01
069775 reg 0x:09 = 0xFF
069775 reg 0x:0A = 0x01
069775 reg 0x:0B = 0xFF
069775 reg 0x:0C = 0x00
069775 reg 0x:0D = 0x00
069775 reg 0x:0E = 0x00
069775 reg 0x:0F = 0x00
069775 reg 0x:10 = 0x00
069775 reg 0x:11 = 0x05
069775 reg 0x:12 = 0x19
069775 reg 0x:13 = 0xFF
069775 reg 0x:14 = 0x19
069775 reg 0x:15 = 0xFF
069775 reg 0x:16 = 0x3E
069775 reg 0x:17 = 0x89
069775 reg 0x:18 = 0x2C
069775 reg 0x:19 = 0x49
069775 reg 0x:1A = 0xB7
069775 reg 0x:1B = 0x93
069775 reg 0x:1C = 0x2E
069775 reg 0x:1D = 0x00
069775 reg 0x:1E = 0x20
069775 reg 0x:1F = 0x00
069776 reg 0x:20 = 0x33
069776 reg 0x:21 = 0x6F
069776 reg 0x:22 = 0x3D
069776 compat> 
```
you can also peek and poke individual registers, ill manually take it out of standby
```
074759 compat> drv2605 poke 1 0
drv2605 poke 1 0
083307 Wrote: 0x00 to 0x01
083307 compat> drv2605 peek 1
drv2605 peek 1
083883 Value: 0x00
```

**Comment by jacobrosenthal** [comment]
ping @microbuilder Been a weeks since I fiddled with it, Im ready to call it ready for review

**Comment by microbuilder** [comment]
@jacobrosenthal I'll try to test this again tonight or tomorrow. Sorry for the delay, the review on on personal time, but I'm interested in getting this chip into the codebase! It's a nice IC and you can do a lot of interesting things with it.

**Comment by microbuilder** [comment]
Finally found a moment to pull a breakout out, and this seems good to me, and the shell options are quite comprehensive which is nice, especially for calibration, etc.!

Do you think a `readme.md` file in `hw/drivers/drv2605` would make sense though, documenting some of the config options there, and including some of the useful code snippets present in this PR, which will (sort of) disappear once it's merged in? That way anyone who browses to that folder on Github will see a well formatted guide to how to use this useful driver in the real world?

**Comment by jacobrosenthal** [comment]
Seems like tutorials in the docs repo are the common way of doing this atm.
Ill get something baed on these instructions over there next

File: hw/drivers/drv2605/syscfg.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    DRV2605_SHELL_ITF_NUM:
+        description: 'DRV2605 interface number'
+        value: 0
+    DRV2605_SHELL_ITF_ADDR:
+        description: 'DRV2605 I2C Address'
+        value: 0x5A
+
+    DRV2605_EN_PIN:
+        description: 'EN pin'
+        value: 3
```
[the remark this answers is not in the document]
Comment (jacobrosenthal) [reply hw/drivers/drv2605/syscfg.yml]: Roger, incoming

File: hw/drivers/drv2605/src/drv2605.c [hunk 2]
```diff
@@ -0,0 +1,872 @@
...
+                               0, 0, 0, 0, 0, 0, 0, 0,
+                               0, 0, 0, 0};
+
+    struct hal_i2c_master_data data_struct = {
+        .address = itf->si_addr,
+        .len = len + 1,
+        .buffer = payload
+    };
+
+    memcpy(&payload[1], buffer, len);
```
[the remark this answers is not in the document]
Comment (jacobrosenthal) [reply hw/drivers/drv2605/src/drv2605.c]: Hah. I actually copied this implementation from 5 other upstream drivers :) Lets merge and update them all.

Is this sufficient?
```
    if (len > 8) {
        goto error;
    }
```

[the remark this answers is not in the document]
Comment (jacobrosenthal) [reply hw/drivers/drv2605/src/drv2605.c]: Oh and which error code

**Comment by aditihilbert** [comment]
@jacobrosenthal , new tutorials are going into https://github.com/apache/mynewt-documentation/tree/master/docs/tutorials instead of the docs repo in core. Keeping core light!

File: hw/drivers/drv2605/src/drv2605.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mkiiskila) [reply hw/drivers/drv2605/src/drv2605.c]: I would've used sizeof(payload) - 1, but this is ok too. All approved from my side.

**Comment by jacobrosenthal** [comment]
@mkiiskila check again. I noticed there are other _writelen with different style. Preferences on how to correct these others
https://github.com/apache/mynewt-core/blob/f7ee032db6b0c3b3e20cac2b7d8ae8d14d33ca71/hw/drivers/sensors/bmp280/src/bmp280.c#L802
https://github.com/apache/mynewt-core/blob/41613e9867edc263166d002c58af3df6717bfc90/hw/drivers/sensors/bme280/src/bme280.c#L886
https://github.com/apache/mynewt-core/blob/f7ee032db6b0c3b3e20cac2b7d8ae8d14d33ca71/hw/drivers/sensors/ms5837/src/ms5837.c#L327

**Comment by mkiiskila** [comment]
Those could be done in a separate PR. IMHO, for static functions the length check does not seem as crucial, as the callers are local to file, and these can easily be eyeballed for incorrect use.

**Comment by jacobrosenthal** [comment]
Sounds good to me, somebody please squash and merge :)

**Comment by jacobrosenthal** [comment]
Since this is approved I figured I can do it myself? Also of note, WIP documentation is being worked on here
https://github.com/apache/mynewt-documentation/pull/25

**Comment by mkiiskila** [comment]
You can merge your PRs yourself.

We seem to have adopted a method of doing this where the person who opened the PR merges the code in after it has been reviewed. Obviously if the PR was opened by someone who is not in the committer list then we have to merge it on their behalf.

**Pull request merged**